### ISSUE TITLE
BYOL-A is incompletely implemented...

### ISSUE THREAD

**Issue by daisukelab** [s3prl/s3prl#341]
Dear @leo19941227 or somebody who can help,

I appreciate your interest. I'm the author of the BYOL-A paper.
I got to know that our BYOL-A is implemented as one of the upstream models, but as far as I can find that it has some problems.
https://github.com/s3prl/s3prl/tree/master/s3prl/upstream/byol_a

1. The input mel feature has to be normalized with the downstream dataset statistics of mean and std. But the current implementation seems to have the **fixed** numbers. This won't work correctly. BYOL-A expects that the input data follows N(0, 1).
2. BYOL-A accepts arbitrary length, but ... implemented to take mel of [64, 96]. And the forward_in_chunks processes chunked input. It can be fixed to process the entire input at once.

I think I can help fix issues and make a pull request, but I have trouble with the upstream interface design regarding the above-mentioned issues.

Is it possible to find any information about:
Question #1: How can we pre-compute the input dataset statistics? We could compute per-batch statistics, but it is not accurate. It is best if we could do that in advance.
Question #2: What is the specification for the get_downsample_rates() and forward()? I especially need to know if the "hidden_states" returning feature of the forward() can be as fine a time frame as 10 ms?

**Comment by andi611** [comment]
> Dear @leo19941227 or somebody who can help,
> 
> I appreciate your interest. I'm the author of the BYOL-A paper. I got to know that our BYOL-A is implemented as one of the upstream models, but as far as I can find that it has some problems. https://github.com/s3prl/s3prl/tree/master/s3prl/upstream/byol_a

Hello! Yes, the current implementation is not complete. We would love to have the author's contribution to help us complete the official implementation! Feel free to **modify/replace anything** you need as long as it follows the interface!

> Is it possible to find any information about: Question #1: How can we pre-compute the input dataset statistics? We could compute per-batch statistics, but it is not accurate. It is best if we could do that in advance. 

One way is to implement this in you model's init. For example, the input dataset path is required in your model's init function and you pre-compute it there. We can add another option, where if the dataset path is not given during init, then the `forward()` function will then apply per-batch statistics? Do you think this is a possible workaround?

For example here:
https://github.com/s3prl/s3prl/blob/e063480425372a75fc0c0fc05dd2f9a4336bc2e7/s3prl/upstream/mockingjay/expert.py#L26
You can add your own argument like `options_config`, for example an argument named `compute_dataset_statistics`.

> Question #2: What is the specification for the get_downsample_rates() and forward()? I especially need to know if the "hidden_states" returning feature of the forward() can be as fine a time frame as 10 ms?

`get_downsample_rates()` simply returns a int, you can precompute this according to your model definition and hard-code it in. For example:
https://github.com/s3prl/s3prl/blob/e063480425372a75fc0c0fc05dd2f9a4336bc2e7/s3prl/upstream/mockingjay/expert.py#L57-L58 or https://github.com/s3prl/s3prl/blob/e063480425372a75fc0c0fc05dd2f9a4336bc2e7/s3prl/upstream/wav2vec2/expert.py#L56-L57

The `forward()` method takes a list of waves as input, and returns a dict:
```python
def forward(self, wavs: List[Tensor]) -> Dict[str, Union[Tensor, List[Tensor]]]:
```
So for example, the function takes `[wav1, wav2, wav3, ...]` as input and returns the following:
https://github.com/s3prl/s3prl/blob/e063480425372a75fc0c0fc05dd2f9a4336bc2e7/s3prl/upstream/mockingjay/expert.py#L60-L65

> returning feature of the forward() can be as fine a time frame as 10 ms

The returning features of `forward()` should correspond to the **entire** input wav. However, you can process the input wav in chunks within the `forward()` function (of any size you wish, for example 10ms) and return the concatenated hidden states.
Just make sure `forward()` returns the entirety of the input wav.

If you have any further questions or if I misunderstand anything please let me know ;)

Andy

**Comment by daisukelab** [comment]
Hello Andy, thanks for your help!

> One way is to implement this in you model's init. For example, the input dataset path is required in your model's init function and you pre-compute it there.

It would be nice. Then I'm trying to find how to pass the path from the command line, for example:

    python run_downstream.py -m train -n Test -u byol_a -d voxceleb1 -o config.runner.datapath=/data/VoxCeleb1

So far it does not bring the overridden option `config.runner.datapath` to the model `__init__`. 
Is there a way we can pass options from the command line to the model?
(I'm checking `downstream/runner.py`, but it looks like no way for the upstream model so far...)

> `get_downsample_rates()` simply returns a int, you can precompute this according to your model definition and hard-code it

For me, `downsample_rate` sounds like the s3prl runner will automatically downsample the input data samples, but it would not happen. Instead, can I understand as follows?
- `get_downsample_rates()` is an interface that tells the runner the time length corresponding to the frame of the hidden state (or embeddings) the model outputs.
- The time length is calculated based on 16000 Hz; for example, it can be 160 if embeddings are output to "hidden_states" every 10 ms.

Thanks, Daisuke

**Comment by leo19941227** [comment]
Hello @daisukelab and @andi611,

Thanks @andi611 for the detailed explanation!

Yes, you are correct about the `get_downsample_rates`.
This is just an information for the DownstreamExpert to prepare some training materials.
Like Speech Enhancement needs the STFT in the same stride to be masked for the IRM approach; Speaker Diarization needs the speaker labels for every timestamp, and the DownstreamExpert needs the stride information to annotate each timestamp given the groundtruth labels in seconds.

You are also right about the override options for the upstream model preparation.
We have supported this in an internal branch but not yet make it public available.
I will make this adjustment available tomorrow and thanks for letting us know your needs!

Leo

**Comment by daisukelab** [comment]
Hello @leo19941227,

It would be nice if the model could receive information about the downstream task in `model.__init__`.
Or, any additional interface to let the model know before training/test starts, for example `if hasattr(model, 'prepare_for_downstream'): model.prepare_for_downstream(downsteam info)` would be great.
The latter would be safe w.r.t. compatibility.

**Comment by andi611** [comment]
> Then I'm trying to find how to pass the path from the command line, for example:
> 
> ```
> python run_downstream.py -m train -n Test -u byol_a -d voxceleb1 -o config.runner.datapath=/data/VoxCeleb1
> ```
> 
> So far it does not bring the overridden option `config.runner.datapath` to the model `__init__`. Is there a way we can pass options from the command line to the model? (I'm checking `downstream/runner.py`, but it looks like no way for the upstream model so far...)

I think we can pass options from the command line to the upstream model through this argument:
https://github.com/s3prl/s3prl/blob/e063480425372a75fc0c0fc05dd2f9a4336bc2e7/s3prl/run_downstream.py#L67

@leo19941227 can you verify if this is correct?
Thanks~

Andy

**Comment by daisukelab** [comment]
@andi611 Hi Andy, thanks!
You are correct if we want to switch file-based configuration.
Though this already allows us to on-the-fly configuration in the command line, we would need to prepare config files for each dataset.
Then, it could be better to pass the name or path of the downstream task directly to the upstream model...

**Comment by andi611** [comment]
> --upstream_model_config

Yes, but `--upstream_model_config` is essentially an argument that also takes a path, 
so you could "pass the name or path of the downstream task" through `-g /path/to/dataset`.

Is the `-g` argument currently being used by your model init or still available?

**Comment by leo19941227** [comment]
> Hello @leo19941227,
> 
> Thank you for your help. This is just a quick idea: It would be nice if the model could receive information about the downstream task in `model.__init__`. Or, any additional interface to let the model know before training/test starts, for example `if hasattr(model, 'prepare_for_downstream'): model.prepare_for_downstream(downsteam info)` would be great. (The latter might be safe w.r.t. compatibility.)
> 
> Anyway, thanks! Daisuke

Basically I am following @daisukelab 's suggestion, I think the later solution of `prepare_for_downstream` is quite nice. We just need to further discuss about what to pass as the `downstream info`

**Comment by daisukelab** [comment]
Dear @leo19941227,

I have prepared a fix for this issue, then I'd like to make a pull request.
However, I cannot make it to the main branch and am not sure how to proceed.

Here's the request I once closed.

https://github.com/s3prl/s3prl/pull/463

Thanks!
